addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py

```diff
@@ -0,0 +1,156 @@
+        """
+        protocol = cls._protocol
+        try:
+            protocol = cls._protocol
+            with serial_connection(device['identifier'], protocol) as connection:
+                connection.write(b'\x09')
+                time.sleep(protocol.commandDelay)
+                response = connection.read(1)
+                if response == b'\x40':
+                    return True
+
+        except serial.serialutil.SerialTimeoutException:
+            pass
+        except Exception:
+            _logger.exception('Error while probing %s with protocol %s' % (device, protocol.name))
+
+    # ----------------
+    # HELPERS
+    # ----------------
+
+    @staticmethod
+    def generate_checksum(message):
+        """ Generate the checksum bytes for the bytes provided.
+
+        :param message: bytes representing the part of the message from which the checksum is calculated
+        :returns:       two checksum bytes calculated from the message
+
+         This checksum is calculated as:
+        1) XOR of all bytes of the bytes
+        2) Conversion into the two bytes of the checksum by adding 30h to each
+           character of the XOR
+
+        eg. to_check = \x12\x23\x34\x45\x56
+            XOR of all bytes in to_check = \x16
+            checksum generated as \x16 -> \x31 \x36
+        """
+        xor = reduce(lambda a, b: a ^ b, message)
+        return bytes([(xor >> 4) + 0x30, (xor & 0xf) + 0x30, ])
+
+    # ----------------
+    # COMMUNICATION
+    # ----------------
+
+    def send(self, msgs):
+        """ Send and receive messages to/from the fiscal device over serial connection
+
+        Generate the wrapped message from the msgs and send them to the device.
+        The wrapping contains the <STX> (starting byte) <LEN> (length byte)
+        and <NBL> (message number byte) at the start and two <CS> (checksum
+        bytes), and the <ETX> line-feed byte at the end.
+        :param msgs: A list of byte strings representing the <CMD> and <DATA>
+                     components of the serial message.
+        :return:     A list of the responses (if any) from the device. If the
+                     response is an ack, it wont be part of this list.
+        """
+
+        replies = []
+        for msg in msgs:
+            self.message_number += 1
+            core_message = struct.pack('BB%ds' % (len(msg)), len(msg)+34, self.message_number+32, msg)
+            request = struct.pack('B%ds2sB' % (len(core_message)), 2, core_message, self.generate_checksum(core_message), 10)
+            with self._device_lock:
+                time.sleep(self._protocol.commandDelay)
+                self._connection.write(request)
+                time.sleep(self._protocol.measureDelay)
+                response = self._connection.read_all()
+            if response[0] == 0x06:
+                if response[2] != 0x30 and response[3] != 0x30:
+                    raise Exception('Error %s %s', response[2], response[3])
```

Comment:
Maybe you can return the human readable error ?
convert this array to a json
`1.2.1. ACKNOWLEDGEMENT RESPONSE`
Like: https://github.com/odoo/enterprise/blob/16.0/pos_l10n_se/iot_handlers/drivers/SerialPosBlackBoxSweden.py#L43